gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaWrapper.java

```diff
@@ -0,0 +1,462 @@
+  private final KafkaAPI kafkaAPI;
+
+  private final boolean useNewKafkaAPI;
+
+  private static class Builder {
+    private boolean useNewKafkaAPI = DEFAULT_USE_NEW_KAFKA_API;
+    private List<String> brokers = Lists.newArrayList();
+
+    private Builder withNewKafkaAPI() {
+      this.useNewKafkaAPI = true;
+      return this;
+    }
+
+    private Builder withBrokers(List<String> brokers) {
+      for (String broker : brokers) {
+        Preconditions.checkArgument(broker.matches(".+:\\d+"),
+            String.format("Invalid broker: %s. Must be in the format of address:port.", broker));
+      }
+      this.brokers = Lists.newArrayList(brokers);
+      return this;
+    }
+
+    private KafkaWrapper build() {
+      Preconditions.checkArgument(!brokers.isEmpty(), "Need to specify at least one Kafka broker.");
+      return new KafkaWrapper(this);
+    }
+  }
+
+  private KafkaWrapper(Builder builder) {
+    this.useNewKafkaAPI = builder.useNewKafkaAPI;
+    this.brokers = builder.brokers;
+    this.kafkaAPI = getKafkaAPI();
+  }
+
+  /**
+   * Create a KafkaWrapper based on the given type of Kafka API and list of Kafka brokers.
+   *
+   * @param state A {@link State} object that should contain a list of comma separated Kafka brokers.
+   * It may optionally specify whether to use the new Kafka API.
+   */
+  public static KafkaWrapper create(State state) {
+    Preconditions.checkNotNull(state.getProp(KAFKA_BROKERS), "Need to specify at least one Kafka broker.");
+    KafkaWrapper.Builder builder = new KafkaWrapper.Builder();
+    if (state.getPropAsBoolean(USE_NEW_KAFKA_API, DEFAULT_USE_NEW_KAFKA_API)) {
+      builder = builder.withNewKafkaAPI();
+    }
+    return builder.withBrokers(state.getPropAsList(KAFKA_BROKERS)).build();
+  }
+
+  public List<String> getBrokers() {
+    return this.brokers;
+  }
+
+  public List<KafkaTopic> getFilteredTopics(Set<String> blacklist, Set<String> whitelist) {
+    return this.kafkaAPI.getFilteredTopics(blacklist, whitelist);
+  }
+
+  public long getEarliestOffset(KafkaPartition partition) throws KafkaOffsetRetrievalFailureException {
+    return this.kafkaAPI.getEarliestOffset(partition);
+  }
+
+  public long getLatestOffset(KafkaPartition partition) throws KafkaOffsetRetrievalFailureException {
+    return this.kafkaAPI.getLatestOffset(partition);
+  }
+
+  public Iterator<MessageAndOffset> fetchNextMessageBuffer(KafkaPartition partition, long nextOffset, long maxOffset) {
+    return this.kafkaAPI.fetchNextMessageBuffer(partition, nextOffset, maxOffset);
+  }
+
+  private KafkaAPI getKafkaAPI() {
+    if (this.useNewKafkaAPI) {
+      return new KafkaNewAPI();
+    } else {
+      return new KafkaOldAPI();
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    this.kafkaAPI.close();
+  }
+
+  private abstract class KafkaAPI implements Closeable {
+    protected abstract List<KafkaTopic> getFilteredTopics(Set<String> blacklist, Set<String> whitelist);
```

Comment:
Would there actually be a difference in how this method would be implemented for the new vs. the old kafka api?